gnomad_qc/v5/data_ingestion/federated_validity_checks.py

```diff
@@ -14,6 +20,97 @@
+    # Determine missingness of each struct annotation.
+    metric_missingness = {}
+    for metric in struct_annotations:
+        metric_missingness.update(check_missingness_of_struct(ht[metric], metric))
+
+    missingness_struct = ht.aggregate(hl.struct(**metric_missingness))
+    missingness_dict = flatten_missingness_struct(missingness_struct)
+
+    logger.info("Checking for missingness within indexed array annotations...")
+    # Determine missingness of each indexed array annotation.
+    # missingness_dict.update(check_array_struct_missingness(ht))
+    missingness_dict.update(check_array_struct_missingness(ht))
+
+    # Report whether or not each metric pass or fails the missingness check
+    # based on the missingness_threshold.
+    for field, missingness in missingness_dict.items():
+        if missingness > missingness_threshold:
+            logger.info(
+                "FAILED missingness check for %s: %.2f%% missing",
+                field,
+                100 * missingness,
+            )
+        else:
+            logger.info(
+                "Passed missingness check for %s: %.2f%% missing",
+                field,
+                100 * missingness,
+            )
+
+    logger.info("Checking for missingness of info and non-info fields...")
+    # Gather info and non-info metrics (or if doesn't exists, set to an empty list).
+    # substract missingness dict
+    # info_metrics = list(ht.row.info) - missingness_dict.keys() or []
+    # non_info_metrics = list(ht.row).remove("info") - missingness_dict.keys() or []
+    info_metrics = set(ht.row.info) - missingness_dict.keys()
+    non_info_metrics = set(ht.row) - {"info"} - missingness_dict.keys()
+    n_sites = ht.count()
+    logger.info("Info metrics are %s", info_metrics)
+    logger.info("Non-info metrics are %s", non_info_metrics)
+    compute_missingness(
+        ht, info_metrics, non_info_metrics, n_sites, missingness_threshold
+    )
+
+
+def validate_federated_data(ht, missingness_threshold: float = 0.50) -> None:
+    """
+    Perform validity checks on federated data..
+
+    :param ht: Input Table.
+    :param missingness_threshold: Upper cutoff for allowed amount of missingness. Default is 0.50.
+    :param struct_annotations: List of struct annotations to check for missingness. Default is  ['grpmax', 'fafmax', 'histograms'].
+    :return: None
+    """
+
+    # Summarize variant and check that all contig exist.
+    expected_contigs = [
+        i
+        for i in hl.get_reference("GRCh38").contigs
+        if i in [f"chr{i}" for i in range(1, 23)] + ["chrX", "chrY"]
+    ]
+    logger.info("Summarizing variants and checking contigs...")
+    summarize_variants(ht, expected_contigs=expected_contigs)
+
+    # Check for missingness.
+    check_missingness(
+        ht, missingness_threshold, struct_annotations=["grpmax", "fafmax", "histograms"]
+    )
+
+    # TODO: consider adding check_global_and_row_annot_lengths
```

Comment:
we should also check here that they actually provided us with two sets of frequencies (raw and adj)